File: bank-service/src/main/java/rs/edu/raf/mapper/CardMapper.java
package rs.edu.raf.mapper;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.CardDTO;
import rs.edu.raf.dto.CreateCardDto;
import rs.edu.raf.model.card.Card;
import rs.edu.raf.model.card.CardIssuer;
import rs.edu.raf.model.card.CardStatus;
import rs.edu.raf.model.card.CardType;
import rs.edu.raf.repository.CardIssuerRepository;
import rs.edu.raf.repository.accounts.BankAccountRepository;

import java.security.SecureRandom;
import java.time.LocalDate;
import java.time.ZoneId;

@Component
@AllArgsConstructor
public class CardMapper {
    private BankAccountMapper bankAccountMapper;
    private BankAccountRepository bankAccountRepository;
    private SecureRandom secureRandom;
    private CardIssuerRepository cardIssuerRepository;
    private CardIssuerMapper cardIssuerMapper;

    public Card createCartDtoToCart(CreateCardDto createCardDto) {
        Card card = new Card();

        CardIssuer cardIssuer = cardIssuerRepository.findById(createCardDto.issuerId()).orElseThrow();

        card.setType(CardType.valueOf(createCardDto.type()));
        card.setCardIssuer(cardIssuer);
        card.setName(createCardDto.name());
        card.setCreationDate(System.currentTimeMillis());

        long expirationDate = LocalDate.now().plusMonths(60).withDayOfMonth(1)
                .atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli();

        card.setExpirationDate(expirationDate);
        card.setBankAccount(bankAccountRepository.findByAccountNumberAndActiveIsTrue(createCardDto.bankAccountNumber()).orElseThrow());
        card.setCvv(secureRandom.nextInt(900) + 100);
        card.setCardLimit(createCardDto.limit());
        card.setCardStatus(CardStatus.ACTIVE);
        card.setNumber(generateCardNumber(card));

        return card;
    }

    public CardDTO cardToCardDto(Card card) {
        return new CardDTO(card.getId(), card.getNumber(), card.getType().toString(),
                cardIssuerMapper.cardIssuerToCardIssuerDTO(card.getCardIssuer()), card.getName(), card.getCreationDate(),
                card.getExpirationDate(), bankAccountMapper.bankAccountToBankAccountDTO(card.getBankAccount()),
                card.getCvv(), card.getCardLimit(), card.getCardStatus().toString());
    }

    private String generateCardNumber(Card card){
        CardIssuer cardIssuer = card.getCardIssuer();
        StringBuilder cardNumber = new StringBuilder();
        cardNumber.append(cardIssuer.getMII());
        cardNumber.append(cardIssuer.getBIN());

        String bankAccountNumber = card.getBankAccount().getAccountNumber().toString();
        cardNumber.append(bankAccountNumber, bankAccountNumber.length() - 10, bankAccountNumber.length() - 2);

        int checkDigit = calculateLuhnCheckDigit(cardNumber.toString());
        cardNumber.append(checkDigit);

        return cardNumber.toString();
    }

    private int calculateLuhnCheckDigit(String number){
        int sum = 0;
        boolean alternate = false;

        for (int i = number.length() - 1; i >= 0; i--) {
            int n = Character.getNumericValue(number.charAt(i));
            if (alternate) {
                n *= 2;
                if (n > 9) {
                    n -= 9;
                }
            }
            sum += n;
            alternate = !alternate;
        }

        return (10 - (sum % 10)) % 10;
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/mapper/CardMapper.kt
package org.example.bankservicekotlin.mapper

import org.example.bankservicekotlin.model.card.Card
import org.example.bankservicekotlin.model.card.CardType
import org.example.bankservicekotlin.model.card.CardStatus
import org.example.bankservicekotlin.model.card.CardIssuer
import org.example.bankservicekotlin.dto.CreateCardDTO
import org.example.bankservicekotlin.dto.CardDTO
import org.example.bankservicekotlin.repository.CardIssuerRepository
import org.example.bankservicekotlin.repository.accounts.BankAccountRepository
import org.springframework.stereotype.Component
import java.security.SecureRandom
import java.time.LocalDate
import java.time.ZoneId

@Component
class CardMapper(
    private val bankAccountMapper: BankAccountMapper,
    private val bankAccountRepository: BankAccountRepository,
    private val secureRandom: SecureRandom,
    private val cardIssuerRepository: CardIssuerRepository,
    private val cardIssuerMapper: CardIssuerMapper
) {

    fun createCartDtoToCart(createCardDto: CreateCardDTO): Card {
        val cardIssuer: CardIssuer = cardIssuerRepository.findById(createCardDto.issuerId).orElseThrow()
        val cardType = CardType.valueOf(createCardDto.type)
        val number = generateCardNumber(cardIssuer, createCardDto.bankAccountNumber.toString())
        val expirationDate = LocalDate.now().plusMonths(60).withDayOfMonth(1)
            .atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli()
        val bankAccount =
            bankAccountRepository.findByAccountNumberAndActiveIsTrue(createCardDto.bankAccountNumber).orElseThrow()
        val cvv = secureRandom.nextInt(900) + 100
        return Card(
            null,
            number,
            cardType,
            cardIssuer,
            createCardDto.name,
            System.currentTimeMillis(),
            expirationDate,
            bankAccount,
            cvv,
            createCardDto.limit,
            CardStatus.ACTIVE
        )
    }

    fun cardToCardDto(card: Card): CardDTO {
        return CardDTO(
            card.id, card.number, card.type.toString(),
            cardIssuerMapper.cardIssuerToCardIssuerDTO(card.cardIssuer), card.name, card.creationDate,
            card.expirationDate, bankAccountMapper.bankAccountToBankAccountDTO(card.bankAccount),
            card.cvv, card.cardLimit, card.cardStatus.toString()
        )
    }

    private fun generateCardNumber(cardIssuer: CardIssuer, bankAccountNumber: String): String {
        val cardNumber = StringBuilder()
        cardNumber.append(cardIssuer.MII)
        cardNumber.append(cardIssuer.BIN)

        cardNumber.append(bankAccountNumber, bankAccountNumber.length - 10, bankAccountNumber.length - 2)

        val checkDigit = calculateLuhnCheckDigit(cardNumber.toString())
        cardNumber.append(checkDigit)

        return cardNumber.toString()
    }

    private fun calculateLuhnCheckDigit(number: String): Int {
        var sum = 0
        var alternate = false

        for (i in number.length - 1 downTo 0) {
            var n = Character.getNumericValue(number[i])
            if (alternate) {
                n *= 2
                if (n > 9) {
                    n -= 9
                }
            }
            sum += n
            alternate = !alternate
        }

        return (10 - (sum % 10)) % 10
    }
}